src/handshake/client.rs

```diff
@@ -49,7 +49,12 @@ impl<S: Read + Write> ClientHandshake<S> {
         // Check the URI scheme: only ws or wss are supported
         let _ = crate::client::uri_mode(request.uri())?;
 
-        let key = generate_key();
+        let key = if let Some(key) = request.headers_mut().remove("Sec-WebSocket-Key") {
+            key.to_str()?.to_owned()
+        }
+        else {
+            generate_key()
+        };
```

Comment:
This could be rewritten as `request.headers_mut().remove("Sec-WebSocket-Key").map(|key| ...).unwrap_or_else(|| generate_key())`.